### PR TITLE
storage: don't wrap single querier in merge-queriers

### PR DESCRIPTION
**Pull request by bboreham**
If given a single querier, just return it instead of constructing a complicated wrapper. The code in `mergeGenericQuerier` which skipped merging when there was only one is not needed any more.

This change required a few tests to be tweaked, because they relied on the specific behaviour of `mergeGenericQuerier.Select()`.

Change suggested by @colega in #13427

### CODE REVIEW

File: storage/merge.go [hunk 1]
```diff
@@ -77,6 +83,15 @@ func NewMergeQuerier(primaries, secondaries []Querier, mergeFn VerticalSeriesMer
 // In case of overlaps between the data given by primaries' and secondaries' Selects, merge function will be used.
 // TODO(bwplotka): Currently merge will compact overlapping chunks with bigger chunk, without limit. Split it: https://github.com/prometheus/tsdb/issues/670
 func NewMergeChunkQuerier(primaries, secondaries []ChunkQuerier, mergeFn VerticalChunkSeriesMergeFunc) ChunkQuerier {
+	switch {
+	case len(primaries) == 0 && len(secondaries) == 0:
+		return noopChunkQuerier{}
+	case len(primaries) == 1 && len(secondaries) == 0:
+		return primaries[0]
+	case len(primaries) == 0 && len(secondaries) == 1:
```
[the remark this answers is not in the document]
Comment (machine424) [reply storage/merge.go:91]: With one secondary, I don't see why we'd need concurrentSelect though...

[the remark this answers is not in the document]
Comment (bboreham) [reply storage/merge.go:91]: To clarify: you're saying all selects are concurrent now, if we get past the trivial cases?

[the remark this answers is not in the document]
Comment (bboreham) [reply storage/merge.go:91]: ISTM that `len(primaries) == 2 && len(secondaries) == 0` would be non-concurrent previously, and I need the `len` to maintain that behaviour.

However it seems weird behaviour, and as far as I can see can never be called that way inside Prometheus.

[the remark this answers is not in the document]
Comment (machine424) [reply storage/merge.go:91]: You're right, I misread the conditions :)

**Review by machine424 — APPROVED** [review]
lgtm, thanks!

File: storage/merge.go [hunk 2]
```diff
@@ -45,9 +45,15 @@ type mergeGenericQuerier struct {
...
+		return noopQuerier{}
+	case len(primaries) == 1 && len(secondaries) == 0:
+		return primaries[0]
+	case len(primaries) == 0 && len(secondaries) == 1:
+		return secondaries[0]
 	}
+
 	queriers := make([]genericQuerier, 0, len(primaries)+len(secondaries))
 	for _, q := range primaries {
 		if _, ok := q.(noopQuerier); !ok && q != nil {
```
Comment (pracucci) [inline storage/merge.go:59]: If a querier is a `noopQuerier` then we skip it. After this filter, we may still end up with only 1 querier. With previous version of the code we were circuit breaking in `mergeGenericQuerier.Select()` but now the check has been removed from `mergeGenericQuerier.Select()`.

Comment (bboreham) [reply storage/merge.go:59]: Addressed as part of #14496.

**Pull request merged**